core/src/server/rpc_module.rs

```diff
@@ -782,15 +868,19 @@ impl SubscriptionSink {
 	/// use anyhow::anyhow;
 	///
 	/// let mut m = RpcModule::new(());
-	/// m.register_subscription("sub", "_", "unsub", |params, mut sink, _| {
+	/// m.register_subscription("sub", "_", "unsub", |params, pending, _| {
+	///     let mut sink = pending.accept()?;
 	///     let stream = futures_util::stream::iter(vec![Ok(1_u32), Ok(2), Err("error on the stream")]);
 	///     // This will return send `[Ok(1_u32), Ok(2_u32), Err(Error::SubscriptionClosed))]` to the subscriber
 	///     // because after the `Err(_)` the stream is terminated.
-	///     tokio::spawn(sink.pipe_from_try_stream(stream));
+	///     tokio::spawn(async move {
+	///         // jsonrpsee doesn't send an error notification if the stream was closed unless `close` is called.
+	///         let _ = sink.pipe_from_try_stream(stream).await.map_err(|e| sink.close(e));
+	///     });
 	///     Ok(())
 	/// });
 	/// ```
-	pub async fn pipe_from_try_stream<S, T, E>(mut self, mut stream: S) -> Result<(), Error>
+	pub async fn pipe_from_try_stream<S, T, E>(&mut self, mut stream: S) -> Result<(), Error>
```

Comment:
we should update the docs but I'm not really sure about the API here i.e, return `SubscriptionClosed` as an error.

I was thinking of doing something like `Result<bool, Error>` or `Result<Option<()>, Error` but I found that to be even harder to understand

Reply:
So, I think the stance we're taking is to let the subscription decide how to communicate that it's closed (ie to use an error code/message or some "ok" value)? And if a subscription has already been closed, then we return a `SubscriptionClosed` error if somebody tries to use it again (wonder whether we could be consuming `self` to prevent that?).

Is that true?

If so, I guess the subscription going away is always "Ok(())" (it might be that the "ok" responses have communicated this OK already?), and then any errors that are emmitted are just forwarded as `Error` in the response, so the type signature feels ok to me (if I'm not way off the mark with my understanding :))